### DISCUSSION TITLE
HasMany relationship Retrieve relationship's records vs Eager loading

### DISCUSSION THREAD

**Discussion by chatisk** [staudenmeir/eloquent-json-relations#110] [General]
So, I stumbled upon this package in search of a solution for my eager loading issue.

While with laravel, i could retrieve records from a relationship
```php
# This worked
Invoice::find(1)->invoiceEvents()->get();

# This never brought any result (but didnt fire any error as well)
Invoice::query()->with('invoiceEvents')->get();
```

My `Invoice`  is this.

```php
class Invoice extends Model
{

    public function invoiceEvents(): HasMany
    {
        return $this->hasMany(InvoiceEvent::class, 'metadata->invoice_id', 'id');
    }
```

As soon as I used this package's traits upon both models ,`Invoice` and `InvoiceEvent` , it worked. While I am satisfied with this I would like to know if anyone has a clue, why laravel can't handle the eager load.

Thanks in advance

**Comment by staudenmeir** [comment]
Hi @chatisk,
Interesting question:

> This worked
> Invoice::find(1)->invoiceEvents()->get();

Lazy loading works without the package because the JSON key (`metadata->invoice_id`) only affects the query and Laravel supports the necessary SQL natively. 

> This never brought any result (but didnt fire any error as well)
> Invoice::query()->with('invoiceEvents')->get();

Eager loading doesn't work because the JSON key also affects how the query results are processed in Laravel/PHP and that's what Laravel doesn't support natively. Eager loading requires `$result->getAttribute('metadata->invoice_id')` to work correctly and that's the functionality the package needs to add:
https://github.com/staudenmeir/eloquent-json-relations/blob/master/src/HasJsonRelationships.php#L34

**Reply by chatisk** [reply]
Thank you very much for your immediate response
